### DISCUSSION TITLE
I think the lack of dynamic, client side routes (for statically exported sites) is the biggest missing feature

### DISCUSSION THREAD

**Discussion by miloofcroton** [vercel/next.js#28752] [Ideas]
To clarify, I mean dynamic routes for cases where you don't know the route until client-side runtime. It's the classic form of client-side routing that vanilla React apps have used forever. It's great for making your website more app-like.

At the moment, there are dynamic routes, but they have to be defined at build time. I wish there was an escape hatch for this.

Technical challenges: This has been brought up many times before, and my understanding from the developers is that the static export creates just a bunch of html files that have to have names which match the real route in the browser. What are potential ways around this? Could we simply have an opt-in flag per component that would let it mount a react-router sort of router onto a page (let's say `/posts`) so that all `/posts/:id` routes just load that same html file? Why can't that be done? It's literally what is done with vanilla react apps.

**Comment by miloofcroton** [comment]
This is my current workaround. I'm not happy about it, but it seems to work.

`/pages/posts/[id].tsx`
```
import { useEffect, useState } from 'react';
import { useRouter } from 'next/router';
import Loading from '../../components/Loading';

export default function Post() {
  const [result, setResult] = useState('');
  const router = useRouter();

  useEffect(() => {
    setTimeout(() => {
      if (router.query.id) {
        setResult((router?.query?.id as string).toUpperCase())
      }
    }, 3000);
  }, [router])

  return (
    <div>
      {
        result !== '' ? result : <Loading/>
      }
    </div>
  );
}
```

`pages/_error.tsx`
```
import { useEffect, useState } from 'react'
import { useRouter } from 'next/router'
import Loading from '../components/Loading'
import Error from 'next/error'

const SPA_ROUTES = [
  /\/posts(.)+/,
]


export default function Page() {
  const [err, setErr] = useState(false)
  const router = useRouter()

  const routeIsSPA = SPA_ROUTES
    .find(rte => router.asPath.match(rte))

  if (routeIsSPA) {
    router.push(router.asPath)
  }

  useEffect(() => {
    setTimeout(() => setErr(true), 3000);
    return () => clearTimeout();
  }, [router])

  if (err) {
    return <Error statusCode={404} />
  }
  else {
    return <Loading />
  }
}
```

Some notes:
- What happens (if you reload the page fresh): there is a flash of the error page if you go to `/posts/asdfasdf`, then it redirects to the posts/[id] file, which shows the same loading gif. That then shows until the timeout resolves, and then it shows the `id` uppercased on the page.
- Because the loading gifs are identical, you can't notice there was a transition. I'm thinking of ways to make this better.
- This automatically works with the Next.js server runtime as well as with just a static file export that is served via a file server. 
- The setTimeout in the posts file simulates an API call. That's where you would take the page's `id` param and do something with it, then hydrate the page.
- There are other ways of doing the matching. I went with regex so I could match paths that have multiple params in them.
- The SPA_ROUTES should be in another file probably. Some sort of site config area maybe.
- The error handling could obviously be built out better, to handle stuff other than 404s and to check for errors in a different way. This was just a quick solution that I had without figuring out the exact sequence of things.

Overall, I'm surprised how seamless this is. I still think it should be part of the actual framework though.

**Reply by longzheng** [reply]
I would suggest using `router.replace()` so that the error page does not appear in the history stack.

```javascript
  if (routeIsSPA) {
    router.replace(router.asPath)
  }
```

**Comment by wKovacs64** [comment]
@miloofcroton is this topic the same as what I'm experiencing and describing here? https://github.com/wKovacs64/next-export-routing

**Reply by miloofcroton** [reply]
Yes, it's the same thing. It's exactly the standard kind of react-router routing that you're probably used to.

**Comment by azaghlou** [comment]
I'm honestly surprised this isn't already a thing

**Comment by longzheng** [comment]
I was also surprised to find this wasn't the default behaviour with Next.js, which I expected coming from `react-router-dom`.

For sites that have an infinite/large number of dynamic pages, it's not practically to statically generate all the pages, as the [`fallback: true` config is designed to accommodate](https://nextjs.org/docs/basic-features/data-fetching#fallback-true) but is not supported by `next export`

I was also expecting `fallback: true` to work with statically exported sites so that some pages (e.g. the most popular ones) can be statically generated but all the other not defined routes load data dynamically at runtime.
